src/shared/uart_framing/uart_framing_test.cpp

```diff
@@ -0,0 +1,177 @@
+static std::vector<uint8_t> generateBigByteVectorToDecode()
+{
+    std::vector<uint8_t> data = generateBigByteVectorToEncode();
+    std::replace(data.begin(), data.end(), 0x00, 0xFF);
+    // insert start and end bytes
+    data.insert(data.begin(), START_END_FLAG_BYTE);
+    data.emplace_back(START_END_FLAG_BYTE);
+
+    // insert overhead byte
+    data.insert(data.begin() + 1, 0x01);
+
+    // insert end overhead byte
+    data.insert(data.end() - 2, 0x02);
+
+    return data;
+}
+
+TEST_P(CobsEncodingTest, encode_decode_test)
+{
+    // Check that cobs encodes to the expected value
+    auto encoded = cobsEncoding(std::get<0>(GetParam()));
+    EXPECT_EQ(encoded, std::get<1>(GetParam()));
+    // Check that cobs decodes to the expected value
+    auto decoded = cobsDecoding(encoded);
+    EXPECT_TRUE(decoded.has_value());
+    EXPECT_EQ(*decoded, std::get<0>(GetParam()));
+}
+
+INSTANTIATE_TEST_CASE_P(
+    encode_decode_test, CobsEncodingTest,
+    ::testing::Values(
+        std::make_tuple(std::vector<uint8_t>({0x01}),
+                        std::vector<uint8_t>({START_END_FLAG_BYTE, 0x02, 0x01,
+                                              START_END_FLAG_BYTE})),
+        std::make_tuple(std::vector<uint8_t>({START_END_FLAG_BYTE}),
+                        std::vector<uint8_t>({START_END_FLAG_BYTE, 0x01, 0x01,
+                                              START_END_FLAG_BYTE})),
+        std::make_tuple(std::vector<uint8_t>({0x01, 0x02, 0x03, 0x04}),
+                        std::vector<uint8_t>({START_END_FLAG_BYTE, 0x05, 0x01, 0x02, 0x03,
+                                              0x04, START_END_FLAG_BYTE})),
+        std::make_tuple(std::vector<uint8_t>({START_END_FLAG_BYTE, START_END_FLAG_BYTE,
+                                              START_END_FLAG_BYTE, START_END_FLAG_BYTE}),
+                        std::vector<uint8_t>({START_END_FLAG_BYTE, 0x01, 0x01, 0x01, 0x01,
+                                              0x01, START_END_FLAG_BYTE})),
+        std::make_tuple(generateBigByteVectorToEncode(),
+                        generateBigByteVectorToDecode())));
+
+class CobsDecodingErrorTest : public ::testing::TestWithParam<std::vector<uint8_t>>
+{
+};
+
+TEST_P(CobsDecodingErrorTest, decode_error_tests)
+{
+    auto decoded = cobsDecoding(GetParam());
+    EXPECT_FALSE(decoded.has_value());
+}
+
+INSTANTIATE_TEST_CASE_P(decode_error_tests, CobsDecodingErrorTest,
+                        ::testing::Values(std::vector<uint8_t>({0x01}),
+                                          std::vector<uint8_t>({0x01,
+                                                                START_END_FLAG_BYTE}),
+                                          std::vector<uint8_t>(100, START_END_FLAG_BYTE),
+                                          std::vector<uint8_t>(100, 0xFF)));
+
+class UartFramingTest : public ::testing::Test
+{
+   protected:
+    struct __attribute__((__packed__)) TestMessage
```

Comment:
do we need to add `__attribute__((__packed__))__` for this to work? have you tried protobuf?

Reply:
Also would this lead to complications given that the Nano uses a different architecture? if `__attribute__((__packed__))__` is needed then i would consider using a byte array, which we can get from protobuf

Reply:
adding __attribute__((__packed__))__ makes the messages smaller sometimes but we should be fine to remove it if we do encounter issues with it.